src/expression/evaluation_context.ts

```diff
@@ -29,8 +40,47 @@ class EvaluationContext {
         return this.feature && 'id' in this.feature ? this.feature.id : null;
     }
 
+    geometryDollarType() {
+        return this.feature ?
+            typeof this.feature.type === 'number' ? geometryTypes[this.feature.type] : simpleGeometryType[this.feature.type] :
+            null;
+    }
+
     geometryType() {
-        return this.feature ? typeof this.feature.type === 'number' ? geometryTypes[this.feature.type] : this.feature.type : null;
+        let geometryType = this.feature.type;
+        if (typeof geometryType === 'number') {
+            geometryType = geometryTypes[this.feature.type];
+            if (geometryType !== 'Unknown') {
+                const geom = this.geometry();
+                const len = geom.length;
+                if (len > 1) {
+                    switch (geometryType) {
+                        case 'Point':
+                            geometryType = 'MultiPoint';
```

Comment:
Is it possible to simply return the relevant value? I think it would simply the method a bit...